content/posts/unpublished/working-with-test-containers/post.md

```diff
@@ -0,0 +1,167 @@
+ );
+
+
+ @DynamicPropertySource
+ static void overrideProperties(DynamicPropertyRegistry registry) {
+   registry.add("spring.kafka.bootstrap-servers", kafka::getBootstrapServers);
+ }
+
+
+ @Autowired
+ private EventEmitter<ProductPriceChangedEvent> productPriceChangedEventEventEmitter;
+
+
+ @Autowired
+ private ProductRepository productRepository;
+
+
+ @BeforeEach
+ void setUp() {
+   Product product = new Product(null, "P100", "Product One", BigDecimal.TEN);
+   productRepository.save(product);
+ }
+
+
+ @Test
+ void shouldHandleProductPriceChangedEvent() {
+   ProductPriceChangedEvent event = new ProductPriceChangedEvent();
+   event.withPrice("14.50")
+     .withCode("P100");
+
+
+   productPriceChangedEventEventEmitter.sendMessage(event);
+
+
+   await()
+     .pollInterval(Duration.ofSeconds(3))
+     .atMost(10, SECONDS)
+     .untilAsserted(() -> {
+       Optional<Product> optionalProduct = productRepository.findByCode("P100");
+
+
+       assertThat(optionalProduct).isPresent();
+       log.info("Product {}", optionalProduct.get());
+       assertThat(optionalProduct.get().getCode()).isEqualTo("P100");
+       assertThat(optionalProduct.get().getPrice())
+         .isEqualTo(new BigDecimal("14.50"));
+     });
+ }
+}
+```
+
+
+* `@SpringBootTest` load the complete Spring app context
+* The Testcontainers special JDBC URL exists to spin up MySQL container and configure it as a DataSource with Spring Boot app context
+* Testcontainers JUnit 5 Extension annotations @Testcontainers and @Container annotations spin up a Kafka container and register the bootstrap-servers location using DynamicPropertySource mechanism.
+* Created a Product record in the database before running the test using the @BeforeEach callback method.
+* During the test `EventEmitter` sends messages to Kafka.
+* As Kafka message processing is an asynchronous process, the Awaitility library checks updates for the product price in the database to the expected value or not with an interval of 3 seconds waiting up to 10 seconds. If the message gets consumed and processed within 10 seconds, the test passes; otherwise, it fails.
+* Also, notice that the property `spring.kafka.consumer.auto-offset-reset` has the value `earliest` so that the listener consume the messages even if the sender dispatches the message to the topic before the listener is ready. This setting is helpful for running tests.
+
+## Conclusion
+Testcontainers emerges as a powerful tool for ensuring reliable and robust testing environments. Unlike mocked and in-memory services, 
+Testcontainers offers the advantage of real-world compatibility, accurately reflecting the behavior of external dependencies such as databases, APIs, message queues.
+By utilizing Testcontainers, developers can identify compatibility issues in the development process at its outset, leading to more resilient software deployments, 
+enhance the effectiveness of testing and improve the reliability of software apps.
+
```

Comment:
may be we should consider mentioning the code example https://github.com/Backbase/working-with-testcontainers